src/hooks/resources/device.ts

```diff
@@ -572,3 +585,58 @@ async function checkSupervisorReleaseUpgrades(
 		}
 	}
 }
+
+async function setSupervisorReleaseResource(
+	api: sbvrUtils.PinejsClient,
+	deviceIds: number[],
+	supervisorVersion: string,
+) {
+	// This function _should_ get called with one device (since each supervisor reports itself)
+	const device = (await api.get({
+		resource: 'device',
+		id: deviceIds[0],
+		options: {
+			$expand: ['is_of__device_type'],
+		},
+	})) as AnyObject;
+
+	if (device.supervisor_version != null) {
+		// if the device already has a supervisor_version, just bail.
+		return null;
+	}
+
+	const supervisorRelease = (await api.get({
+		resource: 'supervisor_release',
+		options: {
+			$select: ['id'],
+			$filter: {
+				supervisor_version: `v${supervisorVersion}`,
+				is_for__device_type: {
+					$any: {
+						$alias: 'dt',
+						$expr: {
+							dt: {
+								id: device.is_of__device_type[0].id,
+							},
+						},
+					},
+				},
+			},
+		},
+	})) as AnyObject[];
+
+	if (supervisorRelease == null) {
+		throw new BadRequestError(
+			// TODO: maybe this shouldn't throw (for the oB case with empty supervisor_releases)
```

Comment:
This error has two disparate cases:

1. In oB, I believe we expect no `supervisor_release`s to exist by design (but this is not necessarily always true).
1. In cloud, we want to not allow a device to set itself to a non-existent release (though the mechanism of "not allow" is perhaps vague and needs to be better elucidated).

Perhaps this means that this code is better suited for the closed-source API?

Reply:
There's also the case of testing releases where they probably have a version that doesn't exist right?

Reply:
When @CameronDiver and I discussed, he mentioned he wanted to handle this at the app level (i.e. as a balenaCloud feature), unless I misunderstood

Reply:
Ok so there was a little confusion here, and we've reconciled it. We're not going to throw an error here, and instead just bail on trying to set the `should_be_managed_by__supervisor_release` field.

Reply:
Yeah, I too think that's the only option really, but we do need to consider soon how openBalena is going to have the Supervisor app and releases populated.